### PR TITLE
P1s7

### PR DESCRIPTION
**Pull request by battjt**
6.1.3 and 6.1.7.2,3 updates.

### CODE REVIEW

File: src-test/org/etools/j1939_84/controllers/part01/Part01Step03ControllerTest.java [hunk 1]
```diff
@@ -223,7 +236,10 @@ public void testModulesEmpty() {
         verify(engineSpeedModule).setJ1939(j1939);
 
         verify(mockListener).addOutcome(1, 3, FAIL, "6.1.3.2.a - No ECU reported as an OBD ECU");
-
+        verify(mockListener).addOutcome(1,
+                                        3,
+                                        WARN,
+                                        "6.1.3.3.b Response received from a non-OBD ECU null provides OBD Compliance values of 0h.");
```
Comment (mmschaefer) [inline src-test/org/etools/j1939_84/controllers/part01/Part01Step03ControllerTest.java]: What is the null?  I think you meant it to be ECU name but we don't always know that so, I would remove that and change the message to "6.1.3.3.b - Response received from a non-OBD ECU provided OBD Compliance value of 0h".  Also, standard for this project is to use dash between number and message along with no periods in the messages save for the ones in the number.

File: src-test/org/etools/j1939_84/controllers/part01/Part01Step03ControllerTest.java [hunk 2]
```diff
@@ -287,7 +303,14 @@ public void testRun() {
         verify(engineSpeedModule).setJ1939(j1939);
 
         verify(mockListener).addOutcome(1, 3, FAIL, "6.1.3.2.b - The request for DM5 was NACK'ed");
-
+        verify(mockListener).addOutcome(1,
```
Comment (mmschaefer) [inline src-test/org/etools/j1939_84/controllers/part01/Part01Step03ControllerTest.java:315]: Add dash and remove period, please.

File: src-test/org/etools/j1939_84/controllers/part01/Part01Step07ControllerTest.java [hunk 3]
```diff
@@ -1166,8 +1169,11 @@ public void testObdModuleCvnAllZerosFailure() {
         verify(mockListener).addOutcome(eq(1),
                                         eq(7),
                                         eq(FAIL),
-                                        eq("6.1.7.2.b.iii - Received CVN is all 0x00 from Electrical System (30)"));
-
+                                        eq("6.1.7.2.b.iv - OBD ECU Received CVN is all 0x00 from Electrical System (30)"));
+        verify(mockListener).addOutcome(eq(1),
+                                        eq(7),
+                                        eq(WARN),
```
Comment (mmschaefer) [inline src-test/org/etools/j1939_84/controllers/part01/Part01Step07ControllerTest.java:1173]: Add dashes and remove periods

File: src-test/org/etools/j1939_84/controllers/part01/Part01Step07ControllerTest.java [hunk 4]
```diff
@@ -1987,11 +1993,15 @@ public void testObdModuleCalIdPaddingIncorrectlyFailure() {
...
+                                        eq("6.1.7.2.b.ii - OBD ECU Brakes - Drive axle #1 (13) CAL ID not formatted correctly (contains non-printable ASCII)"));
         verify(mockListener).addOutcome(eq(1),
                                         eq(7),
                                         eq(FAIL),
-                                        eq("6.1.7.2.b.ii - Brakes - Drive axle #1 (13) CAL ID not formatted correctly (padded incorrectly)"));
+                                        eq("6.1.7.2.b.ii - OBD ECU Brakes - Drive axle #1 (13) CAL ID not formatted correctly (padded incorrectly)"));
+        verify(mockListener).addOutcome(eq(1),
+                                        eq(7),
+                                        eq(WARN),
+                                        eq("6.1.7.3.c CAL ID     calids       has 00h is in either the first or fourth bytes."));
```
Comment (mmschaefer) [inline src-test/org/etools/j1939_84/controllers/part01/Part01Step07ControllerTest.java]: Add dash and remove periods.  Why the weird spacing?

File: src/org/etools/j1939_84/controllers/part01/Part01Step03Controller.java [hunk 5]
```diff
@@ -96,6 +98,29 @@ protected void run() throws Throwable {
             addFailure("6.1.3.2.a - No ECU reported as an OBD ECU");
         }
 
+        // 6.1.3.2.c. Fail if a BEV vehicle response provides a value other than 1Bh (27) for OBD compliance.
+        if (getFuelType() == FuelType.BATT_ELEC) {
+            parsedPackets.stream().forEach(p -> {
+                if (p.getOBDCompliance() != 0x1B) {
+                    addFailure("6.1.3.2.c BEV vehicle reports ODB Compliance other than 1Bh (27).");
```
Comment (mmschaefer) [inline src/org/etools/j1939_84/controllers/part01/Part01Step03Controller.java]: Add dash and remove period, please.

File: src/org/etools/j1939_84/controllers/part01/Part01Step03Controller.java [hunk 6]
```diff
@@ -96,6 +98,29 @@ protected void run() throws Throwable {
...
+                }
+            });
+        }
+        // 6.1.3.2.d. Fail if any response from a function 0 device provides OBD Compliance values of 0, 5, FBh, FCh,
+        // FDh, FEh, or FFh.
+        parsedPackets.stream().forEach(p -> {
+            if (getDataRepository().getFunctionZeroAddress() == p.getSourceAddress() && !p.isObd()) {
+                addFailure("6.1.3.2.d. Fail if any response from a function 0 device provides OBD Compliance values of 0, 5, FBh, FCh, FDh, FEh, or FFh.");
```
Comment (mmschaefer) [inline src/org/etools/j1939_84/controllers/part01/Part01Step03Controller.java]: Add dash and remove period, please.

File: src/org/etools/j1939_84/controllers/part01/Part01Step03Controller.java [hunk 7]
```diff
@@ -96,6 +98,29 @@ protected void run() throws Throwable {
...
+            }
+        });
+
+        // 6.1.3.2.e. Fail if a US/CARB vehicle does not provide OBD Compliance values of 13h, 14h, 22h, or 23h.
+        var usCarbCompliance = List.of(0x13, 0x14, 0x22, 0x23);
+        if (getDataRepository().getVehicleInformation().isUsCarb()
+                && parsedPackets.stream().noneMatch(p -> usCarbCompliance.contains((int) p.getOBDCompliance()))) {
+            addFailure("6.1.3.2.e US/CARB vehicle does not provide OBD Compliance values of 13h, 14h, 22h, or 23h.");
```
Comment (mmschaefer) [inline src/org/etools/j1939_84/controllers/part01/Part01Step03Controller.java]: Add dash and remove period, please.

File: src/org/etools/j1939_84/controllers/part01/Part01Step03Controller.java [hunk 8]
```diff
@@ -110,6 +135,18 @@ protected void run() throws Throwable {
             // All the values should be the same
             addWarning("6.1.3.3.a - An ECU responded with a value for OBD Compliance that was not identical to other ECUs");
         }
+
+        // 6.1.3.3.b Warn if any response received from a non-OBD ECU provides OBD Compliance values of 0, FBh, FCh,
+        // FDh, FEh, or FFh.
+        parsedPackets.stream().forEach(p -> {
+            if (!p.isObd() && p.getOBDCompliance() != 0x5) {
+                addWarning(String.format("6.1.3.3.b Response received from a non-OBD ECU %s provides OBD Compliance values of %Xh.",
```
Comment (mmschaefer) [inline src/org/etools/j1939_84/controllers/part01/Part01Step03Controller.java]: Add dash and remove period, please.  Also, looks like you are printing some formatting accidentally.  In addition, this: `p.getOBDCompliance() != 0x5` confuses me.  I would have expected `!p.isObd() && (p.getOBDCompliance() == 0x0 || p.getOBDCompliance() > 0xFA)` to match the requirements more closely and make it more readable in the future.  I know that is probably picky and you most likely know something I don't but that requires a lot more thinking/processing when you come back in a week/month/year/etc....

File: src/org/etools/j1939_84/controllers/part01/Part01Step07Controller.java [hunk 9]
```diff
@@ -205,9 +205,9 @@ protected void run() throws Throwable {
                 if (allZeros) {
                     String moduleName = packet.getModuleName();
                     if (isObdModule(packet.getSourceAddress())) {
-                        addFailure("6.1.7.2.b.iv - Received CVN is all 0x00 from " + moduleName);
+                        addFailure("6.1.7.2.b.iv - OBD ECU Received CVN is all 0x00 from " + moduleName);
                     } else {
-                        addFailure("6.1.7.3.d.v Received CVN that is all 0x00 from " + moduleName);
+                        addWarning("6.1.7.3.d.v Non-OBD ECU Received CVN that is all 0x00 from " + moduleName);
```
Comment (mmschaefer) [inline src/org/etools/j1939_84/controllers/part01/Part01Step07Controller.java]: Add dash, please.

File: src-test/org/etools/j1939_84/controllers/part01/Part01Step03ControllerTest.java [hunk 10]
```diff
@@ -182,6 +184,14 @@ public void testBadECUValue() {
                                         3,
                                         WARN,
                                         "6.1.3.3.a - An ECU responded with a value for OBD Compliance that was not identical to other ECUs");
+        verify(mockListener).addOutcome(1,
+                                        3,
+                                        WARN,
+                                        "6.1.3.3.b Response received from a non-OBD ECU null provides OBD Compliance values of 0h.");
```
Comment (mmschaefer) [inline src-test/org/etools/j1939_84/controllers/part01/Part01Step03ControllerTest.java]: Add dash and remove period, please.

File: src-test/org/etools/j1939_84/controllers/part01/Part01Step03ControllerTest.java [hunk 11]
```diff
@@ -182,6 +184,14 @@ public void testBadECUValue() {
...
+        verify(mockListener).addOutcome(1,
+                                        3,
+                                        FAIL,
+                                        "6.1.3.2.e US/CARB vehicle does not provide OBD Compliance values of 13h, 14h, 22h, or 23h.");
```
Comment (mmschaefer) [inline src-test/org/etools/j1939_84/controllers/part01/Part01Step03ControllerTest.java]: Add dash and remove period, please.

File: src-test/org/etools/j1939_84/controllers/part01/Part01Step03ControllerTest.java [hunk 12]
```diff
@@ -287,7 +303,14 @@ public void testRun() {
...
         verify(mockListener).addOutcome(1, 3, FAIL, "6.1.3.2.b - The request for DM5 was NACK'ed");
-
+        verify(mockListener).addOutcome(1,
+                                        3,
+                                        FAIL,
+                                        "6.1.3.2.d. Fail if any response from a function 0 device provides OBD Compliance values of 0, 5, FBh, FCh, FDh, FEh, or FFh.");
+        verify(mockListener).addOutcome(1,
+                                        3,
+                                        WARN,
+                                        "6.1.3.3.b Response received from a non-OBD ECU null provides OBD Compliance values of 4h.");
```
Comment (mmschaefer) [inline src-test/org/etools/j1939_84/controllers/part01/Part01Step03ControllerTest.java]: Add dash and remove period, please.

File: src-test/org/etools/j1939_84/controllers/part01/Part01Step03ControllerTest.java [hunk 13]
```diff
@@ -287,7 +303,14 @@ public void testRun() {
         verify(engineSpeedModule).setJ1939(j1939);
 
         verify(mockListener).addOutcome(1, 3, FAIL, "6.1.3.2.b - The request for DM5 was NACK'ed");
-
+        verify(mockListener).addOutcome(1,
+                                        3,
+                                        FAIL,
+                                        "6.1.3.2.d. Fail if any response from a function 0 device provides OBD Compliance values of 0, 5, FBh, FCh, FDh, FEh, or FFh.");
```
Comment (mmschaefer) [inline src-test/org/etools/j1939_84/controllers/part01/Part01Step03ControllerTest.java]: Add dash and remove period, please.

File: src-test/org/etools/j1939_84/controllers/part01/Part01Step07ControllerTest.java [hunk 14]
```diff
@@ -425,20 +426,24 @@ public void testRealDataFromTruck() {
                                         eq("6.1.7.3.a - Total number of reported CAL IDs is > user entered value for number of emission or diagnostic critical control units"));
         verify(mockListener).addOutcome(eq(PART_NUMBER),
                                         eq(STEP_NUMBER),
-                                        eq(WARN),
+                                        eq(INFO),
                                         eq("6.1.7.3.b - Engine #1 (0) provided more than one CAL ID and CVN pair in a single DM19 message"));
+        verify(mockListener).addOutcome(eq(PART_NUMBER),
+                                        eq(STEP_NUMBER),
+                                        eq(WARN),
+                                        eq("6.1.7.3.c CAL ID ÿÿÿÿÿÿÿÿÿÿÿÿÿÿÿÿ has 00h is in either the first or fourth bytes."));
```
Comment (mmschaefer) [inline src-test/org/etools/j1939_84/controllers/part01/Part01Step07ControllerTest.java]: Add dash and remove period, please.

File: src-test/org/etools/j1939_84/controllers/part01/Part01Step07ControllerTest.java [hunk 15]
```diff
@@ -740,8 +745,8 @@ public void testNoNacksToGlobalQueryFromObdFailure() {
 
         when(communicationsModule.requestDM19(any(CommunicationsListener.class))).thenReturn(RequestResult.of());
         when(communicationsModule.requestDM19(any(ResultsListener.class), eq(0x00)))
-                                                                .thenReturn(BusResult.of(
-                                                                                         dm19));
+                                                                                    .thenReturn(BusResult.of(
```
Comment (mmschaefer) [inline src-test/org/etools/j1939_84/controllers/part01/Part01Step07ControllerTest.java]: This change probably was checked in accidently.

File: src-test/org/etools/j1939_84/controllers/part01/Part01Step07ControllerTest.java [hunk 16]
```diff
@@ -1075,11 +1078,11 @@ public void testNonObdModuleCvnAllZerosFailure() {
...
                                         eq(7),
                                         eq(WARN),
-                                        eq("6.1.7.3.c.i - Non-OBD ECU Electrical System (30) provided CAL ID"));
+                                        eq("6.1.7.3.d.i - Non-OBD ECU Electrical System (30) provided CAL ID"));
         verify(mockListener).addOutcome(eq(1),
                                         eq(7),
-                                        eq(FAIL),
-                                        eq("6.1.7.3.c.iv Received CVN that is all 0x00 from Electrical System (30)"));
+                                        eq(WARN),
+                                        eq("6.1.7.3.d.v Non-OBD ECU Received CVN that is all 0x00 from Electrical System (30)"));
```
Comment (mmschaefer) [inline src-test/org/etools/j1939_84/controllers/part01/Part01Step07ControllerTest.java]: Add dash, please.

File: src/org/etools/j1939_84/controllers/part01/Part01Step03Controller.java [hunk 17]
```diff
@@ -110,6 +135,18 @@ protected void run() throws Throwable {
...
+                                         p.getModuleName(),
+                                         p.getOBDCompliance()));
+            }
+        });
+
+        // 6.1.3.3.a Info for DM5 replies from non-OBD ECUs.
```
Comment (mmschaefer) [inline src/org/etools/j1939_84/controllers/part01/Part01Step03Controller.java]: Um, 6.1.3.3.b & 6.1.3.3.c appear to be missing.  Did you forget to check those in or am I missing something?

**Review by mmschaefer — CHANGES_REQUESTED** [review]

File: src/org/etools/j1939_84/controllers/part01/Part01Step03Controller.java (hunk 17 above)
Comment (battjt) [reply src/org/etools/j1939_84/controllers/part01/Part01Step03Controller.java]: 6.1.3.3.b is there, but has similar issues.
What does 6.1.3.3.c mean? log info message if a non-obd module responds with DM5 (packet is already logged).

**Pull request merged**